### ISSUE TITLE
RuntimeError: memory access out of bounds - IDBBatchAtomicVFS

### ISSUE THREAD

**Issue by AntonOfTheWoods** [rhashimoto/wa-sqlite#108]
I am using the IDBBatchAtomicVFS in a chrome extension. I can install a quite large DB, and if I only send a few queries to the DB, they return fine. If I use IDBBatchAtomicVFS in a worker on the main website (so not in the chrome ext), then it is *much* slower than the OPFS sync and if I push it to hard, then it sometimes hangs, but I am not getting errors. However, in the chrome extension I can very reliably get these memory errors when I push it. It will start returning queries, then bork with this.

```
wa-sqlite-async-0af0ef34.wasm:0xf0993 Uncaught (in promise) RuntimeError: memory access out of bounds
    at wa-sqlite-async-0af0ef34.wasm:0xf0993
    at Oh.c.<computed> (wa-sqlite-async.mjs:48:321)
    at wa-sqlite-async.mjs:50:183
```
or
```
content.tsx-b9501ab3.js:64 Uncaught (in promise) RuntimeError: memory access out of bounds
    at chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/wa-sqlite-async-0af0ef34.wasm
    at Oh.c.<computed> [as _sqlite3_malloc] (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:109:12452)
    at W1.t.str_new (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:29712)
    at chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:29251
    at AsyncGenerator.next (<anonymous>)
    at n (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:32134)
    at chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:32408
    at Object.th [as getDefinitions] (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:50:94)
    at x (adapter-78aeae99.js:5:1236)
    at c (adapter-78aeae99.js:5:14906)
```
And sometimes this:

```
content.tsx-b9501ab3.js:64 Uncaught (in promise) Error: database disk image is malformed
    at S (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:31769)
    at Object.step (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:29575)
    at async n (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:107:32241)
    at async Object.th [as getDefinitions] (chrome-extension://mliddjnlkinmciggplpkdoiffahabapa/assets/dbs-2d7a9f74.js:50:88)
```

This method (so the query that gets executed against sqlite) has already been called successfully a number of times at this stage but borks after a few.

Alas, there is no sync in chrome extensions, so I'm not sure whether there are any other options I could try. I attempted to use the other IDB VFSes but I couldn't get any of them to get as far as installing my 300MB db. Any pointers much appreciated!

**Comment by rhashimoto** [comment]
Is the query in a read or write transaction? And if it is a write transaction, does the statement that causes the crash actually write or only read?

How are you creating your 300MB database? Can you try `PRAGMA integrity_check` on it to see if that passes before a crash occurs?

If you're able to build wa-sqlite, try building and importing the debug target (`make debug`, import from debug/ instead of dist/). That might give us more informative stack traces.

**Comment by AntonOfTheWoods** [comment]
> Is the query in a read or write transaction? And if it is a write transaction, does the statement that causes the crash actually write or only read?

I currently have an installation phase (run from the options page) and a run phase launched from a content script that communicates with the extension background "service worker" (they aren't actually service workers... the just decided to reuse the name... anyway). The installation phase has only inserts and the run phase only selects. The installation phase has explicit `begin ... commit`s but the run phase has implicit transations via the `select`s

> How are you creating your 300MB database?
SQL create table + bulk inserts. This isn't particularly fast, but it runs fine and I get no errors. Moderate querying of the created database from the content script works fine. It's just when there are lots of queries that this happens.

> Can you try `PRAGMA integrity_check` on it to see if that passes before a crash occurs?
'ok'

**Comment by AntonOfTheWoods** [comment]
> If you're able to build wa-sqlite, try building and importing the debug target (`make debug`, import from debug/ instead of dist/). That might give us more informative stack traces.

The errors aren't exactly the same... but with the debug I am getting:

```
wa-sqlite-async.mjs:601 Aborted(Assertion failed: We cannot start an async operation when one is already flight)
we	@	wa-sqlite-async.mjs:601
L	@	wa-sqlite-async.mjs:317
vt	@	wa-sqlite-async.mjs:5662
(anonymous)	@	wa-sqlite-async.mjs:5686
(anonymous)	@	sqlite-api.js:689
r	@	tag.ts:53
await in r (async)		
(anonymous)	@	tag.ts:73
nl	@	sqldata.ts:152
x	@	comlink.mjs:103
c	@	adapter.ts:115
```

It very much seems to be a concurrency issue - I may just be hitting it too hard. When I just do things bit by bit (basically I am selecting text on a page, which then gets sent to the background script, which looks up all the words and translates, etc. - if I slowly, manually select bits of text then no errors), when I tell it to do lots it fails with this message with the debug version.

**Comment by rhashimoto** [comment]
> `Aborted(Assertion failed: We cannot start an async operation when one is already flight)`

Ah, okay. The problem is that your application is calling in to SQLite while a previous call is suspended (and incomplete). This is [not supported by Asyncify](https://emscripten.org/docs/porting/asyncify.html):

>It is not safe to start an async operation while another is already running. The first must complete before the second begins.
> 
> Such interleaving may also break assumptions in your codebase. For example, if a function uses a global and assumes nothing else can modify it until it returns, but if that function sleeps and an event causes other code to change that global, then bad things can happen.

In practice, this means that you need to sequence calls to make sure that each query finishes before you launch the next one. If queries are posted by external events, e.g. a message or user action, overlapping can be prevented with an explicit queue or by chaining Promises.

**Comment by AntonOfTheWoods** [comment]
> There's an example of queuing queries [here](https://github.com/rhashimoto/wa-sqlite/blob/122c46ea4d2f20308c0a3d62b79add048a2bef9c/demo/ahp-worker.js#L26-L35). This particular queuing is to make sure that the database connection is only used by one query at a time (which is also important) and not because of Asyncify, but the idea is the same.

Yeah, I actually started out with that `query` as inspiration so I didn't get why I need to add a second level... I ended up turning the `DatabaseService` into a `Proxy` (which I also filched from `SharedService`!), and have that `query` chaining in the `#createProxy`, so whenever a proxy method gets called on the database service it runs through the promise queue. At least it does on the web side. I was bypassing that with the extension hence the bug...

It is going to be a while before I get used to the errors... Thanks for your help!
